Upgrade mapper/router to new IR

pablolh opened the pull request:
A long journey through the mapper

- New IR in place of old IR
- New DDG is used in place of old scheduler
- Various code improvements (e.g. return result instead of modify reference parameter, paths as lists)
- FreeCycle stores its map as an association list mapping new IR references to uints
- Had to implement "remaining" path length in new DDG. Seems there is already some sort of implementation of criticality but those require scheduling to run first
- Random number generation is seeded with constant seed

Note: the code is still quite untested. I've done my best to check that the program works as expected, but given the number of options and complexity of the code, only proper unit testing for each part and class (e.g. AlterTest, PastTest) can give a good guarantee of functionality. This would take quite some time to implement.

I have updated the docs to signal when an option is only valid when using the minextend/minextendrc route heuristics.

Commit:
Upgrade mapper/router to new IR

- New IR in place of old IR
- New DDG is used in place of old scheduler
- Various code improvements (e.g. return result instead of modify
  reference parameter)
- Remove BASE_RC and MINEXTEND_RC mapper options
- Really order available gates by decreasing "criticality" which was the
original intent but was gonne from the code.

Commit:
Fix order of iteration of DDG node successors/predecessors

Allow an iteration order that does not depend on where the statements
are allocated in memory. Indeed, utils::One::operator< was used to order
the keys in the map, which can change between runs.

Commit:
Fix failing unitary Python test

pablolh commented:
There is a new test map_test.cc that is not really a unit test but gives some guarantee as to preservation of the semantics of the circuit. It's quite basic though.
There is potentially more work to do on this mapper/router but let's not spend too much time on this since there are more urgent matters. namely removing the old IR

My work ethics tell me to signal that the decomposition to primitives is quite untested since I haven't set mapdecompositionrulepattern in any test (so the old tests now don't make use of the finer-grained scheduling obtained by decomposing to primitives) and I haven't updated the cclight platform configurations to use new-style decompositions.

See also #482

jvansomeren commented on .gitmodules:
```diff
@@ -6,10 +6,6 @@
 	path = deps/eigen
 	url = https://gitlab.com/libeigen/eigen.git
 	shallow = true
-[submodule "deps/backward-cpp"]
```
Why is this out? A nice stack backtrace on a fault is nice!

pablolh replied:
I don't like it because
- stack traces in general don't provide enough information for debugging. You need to look around at variables with gdb to know the root cause of the bug, and if you do that then the backtrace is useless because gdb does it better
- users don't care, only developers do, and developers should rather use gdb
- when enabled, it pollutes the output because it dumps the whole stack trace for a simple exception, so a simple error message is

pablolh replied:
To be continued :D

jvansomeren replied:
Base issue is that a stack trace allows me to start debugging without gdb, e.g. from a test program that is written in Python and avoids running something like clion and knowing how to intercept the execution of the OpenQL C++ module from Python to start gdb for it. Or you must train me ...

pablolh replied:
The stacktrace is back.

jvansomeren commented on include/ql/com/ddg/build.h:
```diff
@@ -21,7 +21,7 @@ class EventGatherer {
     /**
      * Reference to the root of the IR.
```
Comment suggests root reference but this is platform reference.
Why? ir also contains platform doesn't it?

pablolh replied:
**I've updated the comment.**

My original rant (skip reading if not interested):
Indeed the comment is no longer accurate.
That is exactly why my previous employer was telling me not to write comments like this.
They get out of sync and take time and energy to maintain.

jvansomeren commented on include/ql/com/ddg/ops.h:
```diff
@@ -68,6 +68,20 @@ void clear(const ir::BlockBaseRef &block);
+
+/**
+ * Returns whether two ddgs have the same shape and same statements,
+ * will fail when commutation rules are enabled, because a node can then
```
Don't understand why this fails when commutation is enabled. And should it fail then?

pablolh replied:
When commutation is enabled, a given node can have multiple children which are identical instructions
It does fail in this case, there is a QL_ASSERT that fires when a node has equal children

I've moved this function to the mapper test, since it was only used there.

For context: this function tries to check that two graphs are identical/equivalent, and it's a hard problem. This problem is easier to solve when no node has two identical children (based on the value they contain). This function assumes this.

jvansomeren commented on include/ql/com/ddg/types.h:
```diff
@@ -354,7 +354,7 @@ using EdgeCRef = utils::Ptr<const Edge>;
  */
-using Endpoints = utils::Map<ir::StatementRef, EdgeRef>;
+using Endpoints = std::list<std::pair<ir::StatementRef, EdgeRef>>;
 
```
Like to schedule (separate from this review) sessions to understand the modules such as this ddg.

jvansomeren commented on include/ql/com/map/reference_updater.h:
```diff
@@ -0,0 +1,33 @@
+
+    void visit_reference(ir::Reference &ref) override;
+
+    // Gate operands may be virtual qubits, but in the instruction type it's always real qubit indices.
```
As a PR: we need to review the treatment of virtual versus real qubits in the IR. In my opinion whether they are virtual or real or something else (such as remote or local to a core, etc., or virtual in a real core, etc.) should be depending on which passes have run. Then this remark becomes redundant.

pablolh replied:
I get what you're saying.
What I'm saying in the comment only concerns the qubit indices in specialized instruction.
Assume an instruction CNOT is specialized for qubit indices 3 and 4 and then takes 10 cycles instead of 20.
Those indices 3 and 4 are REAL indices, since this concerns the actual topology of the architecture.
So, in **specialized instruction types**, qubit indices are always real.

jvansomeren replied:
Making this an issue is sufficient.

jvansomeren commented on include/ql/com/map/reference_updater.h:
```diff
@@ -0,0 +1,33 @@
+namespace com {
+namespace map {
+
+class ReferenceUpdater : public ir::RecursiveVisitor {
```
What is the use of this?
I'm missing comment on what this does.
Also (or not also) shouldn't this be documented?

pablolh replied:
Added a comment

jvansomeren replied:
I'll update the functional (header) and implementation (.cc) comments and the doc methods meant for readthedocs in a separate pull request after this one.
So all these documentation and comment issues can be collected into an issue, and accepted.

jvansomeren replied:
#479

jvansomeren commented on include/ql/com/sch/scheduler.h:
```diff
@@ -152,7 +152,7 @@ class Scheduler {
 
         // Update the resource state.
-        resource_state->reserve(cycle, statement);
+        resource_state->reserve_new_ir(cycle, statement);
```
Why 'reserve_new_ir' and not just 'reserve'?

pablolh replied:
I had a compilation error because of an ambiguous overload. I should look further how to fix it while keeping the overload.

jvansomeren commented on include/ql/ir/compat/gate.h:
```diff
@@ -70,23 +71,6 @@ std::ostream &operator<<(std::ostream &os, ConditionType condition_type);
-
 /**
  * gate interface
  */
```
This is old IR stuff.
Is it only there to support the old IR API in the new IR?
I'm missing comment and documentation again.

jvansomeren replied:
As far as I understand, the compat stuff will be phased out.
So I propose not to further update it with comments etc.

jvansomeren commented on include/ql/ir/ops.h:
```diff
@@ -108,7 +108,7 @@ InstructionTypeLink add_instruction_type(
  */
 InstructionTypeLink find_instruction_type(
-    const Ref &ir,
+    const PlatformRef &platform,
```
As commented on earlier, why this platform ref  replacing the ir ref. As far as I understand, the platform is part of the IR. Does it only rely on the platform?

pablolh replied:
find_instruction_type should only use the platform no? It for sure doesn't need (programmatically and in spirit) the program.

jvansomeren commented on include/ql/ir/ops.h:
```diff
@@ -316,5 +316,134 @@ utils::UInt get_duration_of_block(const BlockBaseRef &block);
+            QL_FATAL("Operand #" << operandIndex << " of instruction " << instr.instruction_type->name << " is not a reference.");
+        }
+
+        if (ref->target != platform->qubits) {
```
So this is the test to test whether an operand is a qubit.
Are all qubits always in platform->qubits?
Please note the discussion on virtual vs real qubits in this context.

jvansomeren replied:
Should be made part of the issue on IR representation of virtual and real qubits.

jvansomeren replied:
#480

jvansomeren commented on include/ql/ir/ops.h:
```diff
@@ -316,5 +316,134 @@ utils::UInt get_duration_of_block(const BlockBaseRef &block);
+
+        utils::UInt nQubitOperands = 0;
+        for (const auto& op: instr_type.operand_types) {
+            if (op->data_type->type() == NodeType::QubitType) {
```
This is another check on whether an operand is a qubit.
Why?

jvansomeren replied:
Should be made part of the issue on IR representation of virtual and real qubits.

jvansomeren replied:
#480

jvansomeren commented on include/ql/ir/ops.h:
```diff
@@ -316,5 +316,134 @@ utils::UInt get_duration_of_block(const BlockBaseRef &block);
+    bool isNN2QGate(std::function<utils::UInt(utils::UInt)> v2r) {
+        auto qubits = get2QGateOperands();
+
+        return platform->topology->get_min_hops(v2r(qubits.first), v2r(qubits.second)) == 1;
```
This only works when the indices are virtual.

jvansomeren replied:
Should be made part of the issue on IR representation of virtual and real qubits.

jvansomeren replied:
#480

jvansomeren commented on include/ql/ir/swap_parameters.h:
```diff
@@ -0,0 +1,26 @@
+namespace ql {
+namespace ir {
+
+struct SwapParameters {
```
Where is this for?
Is it for the visualizer?
Needs explanation.

jvansomeren replied:
Yes, it is, so ok.

jvansomeren commented on include/ql/rmgr/resource_types/base.h:
```diff
@@ -211,7 +211,7 @@ class Base {
      */
     utils::Bool gate(
-        utils::UInt cycle,
+        utils::Int cycle,
```
Strange, here you changed from Uint to Int and in a later interface, see ql/rmgr/state.h, you changed from Int to UInt.

pablolh replied:
As I see, `ql::rmgr::resource_types::Base::on_gate` which is a virtual function that calls this `gate()` uses signed integers for cycle. So this UInt was wrong wrt the base class and there was an implicit conversion.

jvansomeren commented on include/ql/rmgr/state.h:
```diff
@@ -77,8 +77,8 @@ class State {
      */
-    utils::Bool available(
-        utils::Int cycle,
+    utils::Bool available_new_ir(
```
So here you changed from Int to UInt, while the comment above says that the cycle number may be negative.

pablolh replied:
I think that was a mistake from me
But in spirit I don't think cycle numbers should be negative.
That the scheduler uses that for reserved dependency graph scheduling is an implementation detail that should not leak into rmgr, I think. This is quite dangerous here I think because C++ does implicit conversions that could go wrong...

jvansomeren commented on include/ql/utils/exception.h:
```diff
@@ -111,8 +102,7 @@ class Exception : public std::runtime_error {
      * Creates a new exception object with the given message. If system is set,
      * the standard library is queried for the string representation of the
-     * current value of errno, and this is appended to the message. Finally, a
-     * backtrace is performed and appended to the message after a newline.
```
I would like to have this backtrace back.

jvansomeren replied:
Roberto can decide.

jvansomeren commented on src/ql/com/ddg/dot.cc:
```diff
@@ -59,7 +59,13 @@ void dump_dot(
         os << line_prefix << "  n" << it.first;
-        os << " [ label=<n" << it.first << ", cycle " << it.second->cycle << "<br/>";
+
+        std::string remaining = "";
```
When the consequence of using annotations as alternative to IR extensions is this kind of code, it is becoming ugly. Somehow, you want to access annotations similarly to standard IR. Which are now the interfaces available to access annotations? Perhaps another layer on top of those would do.

jvansomeren replied:
Turn this into an issue: annotations as an alternative to IR extensions, functionality and interfaces.

jvansomeren commented on src/ql/com/ddg/ops.cc:
```diff
@@ -134,6 +135,79 @@ void reverse(const ir::BlockBaseRef &block) {
+        auto current = *toVisit.begin();
+        toVisit.erase(toVisit.begin());
+
+        auto currentRemaining = current->get_annotation<Remaining>().remaining;
```
Again ugly annotation interfacing, blurring easy understanding of the code's intention.

jvansomeren replied:
#481

jvansomeren commented on src/ql/com/ddg/ops.cc:
```diff
@@ -134,6 +135,79 @@ void reverse(const ir::BlockBaseRef &block) {
+            bool found = false;
+            for (const auto& node_edge2: succ2) {
+                if (node_edge2.first->equals(*node_edge1.first)) {
+                    QL_ASSERT(!found); // This fails when commutation rules are enabled
```
why does it fail? could you make it so that it doesn't fail on dogs with commutation enabled?

jvansomeren commented on src/ql/com/dec/rules.cc:
```diff
@@ -20,11 +20,6 @@ namespace dec {
 public:
 
-    /**
-     * Map of variable references that need to be updated.
```
Why were variables in and now out?

pablolh replied:
The new IR instruction decomposition rules has this "objects" thing that is supposed to hold
"Any temporary variables as needed within the decomposition rule."

This is absolutely untested, nor used anywhere.
Moreover the semantics of it is not very precise.
I'm sure Jeroen himself never used it. He probably wrote this code for completeness...

Therefore I considered this to be dead code and likely to be just wrong.

jvansomeren replied:
#482

jvansomeren commented on src/ql/com/map/reference_updater.cc:
```diff
@@ -0,0 +1,44 @@
+        QL_ASSERT(ref.indices.size() == 1);
+        auto& virt = ref.indices[0].as<ir::IntLiteral>()->value;
+        virt = mapping[virt];
+        if (callback) {
```
I couldn't find a definition of callback.
Also its purpose is not clear to me.

pablolh replied:
I added a comment in the header
"c: a callback to be called for every mapped virtual qubit index."

While it might not be the best code, its use is here:
```c++
void Past::make_real(const ir::CustomInstructionRef &gate) {
    const auto &gname = gate->instruction_type->name;
    com::map::ReferenceUpdater::Callback cb = [this, gname](utils::UInt virtual_qubit) {
        if (options->assume_prep_only_initializes && (gname == "prepz" || gname == "Prepz")) {
            v2r.set_state(virtual_qubit, com::map::QubitState::INITIALIZED);
        } else {
            v2r.set_state(virtual_qubit, com::map::QubitState::LIVE);
        }
    };

    com::map::mapInstruction(platform, v2r.get_virt_to_real(), gate, cb);
}
```

That is: it is used to set all operands of a mapped gate as "LIVE"

jvansomeren replied:
I'd like to understand the visitors. Couldn't find them. Are they in ir/ir_gen.h ?
And then on top of those such mapInstruction is defined?
And then ReferenceUpdater?
So the idea is to have a base class with methods that all can be overridden,
and then for a particular visit override some, add a visiting function (such as callback),
and then you have a customized ReferenceUpdater. But why that overriding?
That v2r.get_virt_to_real() is a function reference, I guess; odd that it has () behind it; cb hasn't ().

jvansomeren replied:
#484

jvansomeren commented on src/ql/com/map/reference_updater.cc:
```diff
@@ -0,0 +1,44 @@
+}
+
+void mapInstruction(const ir::PlatformRef &platform, const utils::Vec<utils::UInt> &mapping, const ir::CustomInstructionRef &instr, ReferenceUpdater::Callback callback) {
+    QL_ASSERT(instr->instruction_type->generalization.empty() && "Instruction to map should be in the most generalized form, since it's using virtual qubit indices as operands");
```
Perhaps mapping could start with a 'pass' that checks all preconditions such as this one to not clutter up the code with it.

jvansomeren replied:
Perhaps the mapper/router should just accept specialized AND generalized. It is upto the compiler writer to decide when to specialize, it shouldn't bother mapping/routing. This similar to that the mapper shouldn't be responsible for the schedule.

jvansomeren replied:
Or we could leave this assert out

jvansomeren replied:
I guess I mistakenly considered this part of the mapper (the file name has a 'map' in it).
When this is a general facility, then all 3 comments above should be ignored.

pablolh replied:
It's indeed a "common" function that is used both in the initial placer and mapper/router
It turns an instruction into another instruction, with operands modified as instructed by the input mapping.

jvansomeren replied:
#483

jvansomeren commented on src/ql/com/map/reference_updater.cc:
```diff
@@ -0,0 +1,44 @@
+    ReferenceUpdater visitor(platform, mapping, callback);
+    instr->visit(visitor);
+
+    specialize_instruction(instr);
```
Why specializing it here?

jvansomeren replied:
I guess I mistakenly considered this part of the mapper (the file name has a 'map' in it).
When this is a general facility, then all earlier 3 comments above should be ignored.

pablolh replied:
OK

jvansomeren replied:
#483

jvansomeren commented on src/ql/com/options.cc:
```diff
@@ -362,6 +362,13 @@ Options make_ql_options() {
     );
 
+    options.add_str(
+        "mapdecompositionrulepattern",
```
Very long name not conforming to standard way to form names. Can it be shorter? Otherwise e.g. MapDecompositionRulePattern or map_decomposition_rule_pattern

pablolh replied:
Those are global options, and apparently they all have this style, e.g. above is "mapreverseswap"

jvansomeren replied:
#482

jvansomeren commented on src/ql/com/ddg/build.cc:
```diff
@@ -318,26 +318,32 @@ class Builder {
-        auto result = from.node->successors.insert({to.statement, {}});
-        auto &edge_ref = result.first->second;
-        if (result.second) {
+        auto it = std::find_if(from.node->successors.begin(), from.node->successors.end(),
```
Could this code be hidden as method in the successors implementation?

jvansomeren commented on src/ql/com/ddg/build.cc:
```diff
@@ -318,26 +318,32 @@ class Builder {
+        auto it = std::find_if(from.node->successors.begin(), from.node->successors.end(),
+            [&to](const std::pair<ir::StatementRef, EdgeRef> &x) { return x.first == to.statement; }); // Pointer equality is used!
 
+        if (it == from.node->successors.end()) {
```
Would edge_ref as variable not enhance readability of this code?

pablolh replied:
Created a ref to the list of successors of `from`

jvansomeren commented on src/ql/com/ddg/ops.cc:
```diff
@@ -65,7 +65,8 @@ NodeCRef get_sink_node(const ir::BlockBaseRef &block) {
 EdgeCRef get_edge(const ir::StatementRef &from, const ir::StatementRef &to) {
     const auto &successors = get_node(from)->successors;
-    auto it = successors.find(to);
+    auto it = std::find_if(successors.begin(), successors.end(),
```
See earlier remark (in src/ql/com/ddg/dot.cc?)

jvansomeren commented on src/ql/pass/ana/visualize/detail/types.h:
```diff
@@ -114,7 +114,7 @@ struct GateProperties {
     utils::Vec<utils::Int> operands;
     utils::Vec<utils::Int> creg_operands;
-    ir::compat::SwapParamaters swap_params; // FIXME: change once mapper is moved to new IR.
+    ir::SwapParameters swap_params;
```
Is this all?
Why was there the remark to fix this, what was the problem?

pablolh replied:
SwapParameters is an IR annotation to transfer information from the mapper/router to the visualizer.

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.cc:
```diff
@@ -1,264 +1,89 @@
-    score_valid = false; // will not print score for now
-}
-
-/**
```
Was printing paths taken out?

jvansomeren replied:
I would like the printing to be added everywhere again. And the DOUT triggering not only to depend on DEBUG whatever but also on a class or file static toggle which can be switched from outside dynamically. These outputs are much nicer formatted, compact and so informative than an adb can provide in short time. They helped me to pinpoint bugs quickly. After that you start the real debugging, statement by statement analysis what is happening.

jvansomeren replied:
#485

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.cc:
```diff
@@ -1,264 +1,89 @@
-    total.insert(total.begin(), q); // hopelessly inefficient
-}
-
-/**
```
This comment (after some updates) could be added to add_swaps.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.cc:
```diff
@@ -1,264 +1,89 @@
-            from_source_qubit = to_source_qubit;
-            num_added++;
+        utils::UInt swaps_added = 0;
+        for (auto it = path->begin();
```
Some more comment on what this does, would be nice.

pablolh replied:
done

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.cc:
```diff
@@ -1,264 +1,89 @@
-                past.add_swap(from_source[0], from_source[1]);
+
+        if (leftOpIt != path->begin() && rightOpIt != path->rbegin()) {
+            if (past.is_first_swap_earliest(*path->begin(), *std::next(path->begin()),
```
Giving variables names helps readability.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.cc:
```diff
@@ -1,264 +1,89 @@
-    past.schedule();
 }
 
-/**
```
How the recursion works, how pasts are stacked is useful information and not obvious.
I propose to review this comment and place it again above extend().

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.cc:
```diff
@@ -1,264 +1,89 @@
+    auto rightOpIt = std::prev(std::prev(shared_path->rend()));
+    while (leftOpIt != std::prev(shared_path->end())) {
+        QL_ASSERT(*std::next(leftOpIt) == *rightOpIt);
+        QL_ASSERT(platform->topology->get_distance(*leftOpIt, *rightOpIt) == 1);
```
Original said: assert distance >= 1, so distance can be larger here than 1!
Why was this modified?

pablolh replied:
Distance between leftOpIt and rightOpIt needs to be 1, because it is the hop in the topology where the 2q gate will be executed.
What this loop does is go through the proposed path, and create alters for every hop it goes through.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.cc:
```diff
@@ -1,264 +1,89 @@
 }
 
-/**
- * Split the path. Starting from the representation in the total attribute,
```
All this comment was useful, so please put it back in.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.cc:
```diff
@@ -1,264 +1,89 @@
-        utils::UInt leftopi = rightopi - 1;
-        QL_ASSERT (leftopi >= 0);
-        // QL_DOUT("... leftopi=" << leftopi);
-        // leftopi is the index in total that holds the qubit that becomes the left operand of the gate
```
Again useful comment.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.cc:
```diff
@@ -1,264 +1,89 @@
-        result.push_back(na);
-        // QL_DOUT("... added to result list");
-        // DPRINT("... current alter after split");
+        leftOpIt = std::next(leftOpIt);
```
Don't understand that all this code became so small!
Also the QL_DOUTs are missing, as well as the prints; this is now horrible to debug.

pablolh replied:
The full, unsplit path is stored (and shared) in every resulting alter (so doesn't have to be split per-se), and the partial paths information is stored as 2 pointers to the said full path. Those don't need calculations, since they come directly from iterating the full unsplit path.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.h:
```diff
@@ -21,190 +10,81 @@ namespace qubits {
 namespace map {
 namespace detail {
 
-/**
```
Again, useful comment deleted.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.h:
```diff
@@ -21,190 +10,81 @@ namespace qubits {
-     * Cycle time, multiplier from cycles to nanoseconds.
+     * Add swap gates for the current path to the given past, up to the maximum
+     * specified by the swap selection mode. This past can be a path-local one
+     * or the main past. After having added them, schedule the result into that
```
As far as I understand what you explained what and what not to do, you currently are not scheduling the results into the main past.

Moreover, past is not a member anymore of Alter. So as far as I understand, 'past' use is out-dated here.

pablolh replied:
I think this "past" in the comment refers to the argument to the function add_swaps below.

jvansomeren replied:
#486

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.h:
```diff
@@ -21,190 +10,81 @@ namespace qubits {
 
     /**
-     * The gate that this variation aims to make nearest-neighbor.
+     * Compute cycle extension of the current alternative in curr_past relative
```
Ok, here is the missing comment.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.h:
```diff
@@ -21,190 +10,81 @@ namespace qubits {
 
     /**
-     * Prints a state of a whole list of Alters, prefixed by s.
+     * The gate that this variation aims to make nearest-neighbor.
```
'variation' is the old name for 'alter(native)' left from a few years ago; can you please update the comment?

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.h:
```diff
@@ -21,190 +10,81 @@ namespace qubits {
-     * Adds a node to the path in front, extending its length with one.
-     */
-    void add_to_front(utils::UInt q);
+    std::list<utils::UInt>::iterator leftOpIt;
```
In the path of qubit locations from the location of the first qubit of the target gate to location of the second qubit of the target gate, this leftOpIt points to the location that holds the qubit that will be the first operand of the mapped target gate and rightOpIt points to the second operand of this mapped gate, i.e. the hop between those is implemented by the target gate.
When this is true, I propose to add this kind of text as common there to make all this more understandable.

pablolh replied:
Added comment

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.cc:
```diff
@@ -127,105 +58,79 @@ utils::Bool FreeCycle::is_first_swap_earliest(
-    utils::UInt start_cycle_second_swap = utils::max(fcv[sr0] - 1, fcv[sr1]);
 
-    QL_DOUT("... fcv[" << fr0 << "]=" << fcv[fr0] << " fcv[" << fr1 << "]=" << fcv[fr1] << " start=" << start_cycle_first_swap << " fcv[" << sr0 << "]=" << fcv[sr0] << " fcv[" << sr1 << "]=" << fcv[sr1] << " start=" << start_cycle_second_swap << " is_first_swap_earliest=" << (start_cycle_first_swap < start_cycle_second_swap));
+    auto start_cycle_first_swap = utils::max(get_for_qubit(fr0) - 1, get_for_qubit(fr1));
```
Don't understand this - 1 anymore although it was in the original. Couldn't find my old mapper code to check.
There is no comment explaining this.

Could it be that it assumes silently that a SWAP(c,t) is CNOT(c,t);CNOT(t,c);CNOT(c,t) and a CNOT is Ym90(t);CZ(c,t);Y90(t)
so a SWAP(c,t) starts with an idle cycle for qubit c (fr0 or sr0 here) so the SWAP itself starts 1 cycle earlier than qubit c (fr0 or sr0) starts, and starts exactly when qubit t (fr1 or sr1) starts?

I forgot that this silent assumption was in the code. How to handle this? Kind of introspection of the decomposition of SWAP? But how would the mapper know?
The idea is that a SWAP (and any semantically reflexive 2 qubit gate g(c,t)==g(t,c)) may have an asymmetric execution start of its qubits.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.cc:
```diff
@@ -127,105 +58,79 @@ utils::Bool FreeCycle::is_first_swap_earliest(
  */
-utils::UInt FreeCycle::get_start_cycle_no_rc(const ir::compat::GateRef &g) const {
+utils::UInt FreeCycle::get_start_cycle(const ir::CustomInstructionRef &g) const {
     utils::UInt start_cycle = 1;
```
Is the default start of execution of a block and its gates still cycle 1 and not cycle 0?

pablolh replied:
I don't think it matters, since the mapper will only compare start cycles to choose the best one for routing. So whether it starts at 0 or 1 shouldn't matter?

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.cc:
```diff
@@ -127,105 +58,79 @@ utils::Bool FreeCycle::is_first_swap_earliest(
+    const auto& cond = g->condition;
+    const auto &ref = cond.as<ir::Reference>();
+    if (!ref.empty()) {
+        start_cycle = utils::max(start_cycle, get_for_reference(*ref));
```
Can a condition contain qubit references? What is done here?

pablolh replied:
As far as I can tell, only small parts of OpenQL support control flow e.g. if/else.
I have assumed the mapper only gets an array of gates as input, and it was already the case before.
The only conditions here would be the cond(b[x]) notation, which I believe is taken into account during creation of the DDG, that is, a cond(b[x]) U q[y] would have a dependency on a previous measure q[x]

jvansomeren replied:
So it is this b[x] depending on an earlier q[x] having been measured.

The cycle assignment by the quantum scheduler can be seen as imposing deadlines to the classical scheduler. Those cycles could be (and are probably) transferred from the gates to the microcodes implementing those gates.

#487

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.cc:
```diff
@@ -127,105 +58,79 @@ utils::Bool FreeCycle::is_first_swap_earliest(
-        if (base_start_cycle != start_cycle) {
-            // QL_DOUT(" ... from [" << baseStartCycle << "] to [" << startCycle-1 << "] busy resource(s) for " << g->qasm());
+    if (rs.has_value()) {
+        while (!rs->available_new_ir(start_cycle, g.as<ir::Statement>())) {
```
For understanding of future people it is relevant to explain here that this depends on assumptions of the resource system: once while incrementing the start cycle, all resources needed by all operands of this gate are available, the gate can start at that cycle and can complete its execution, i.e. the resources remain available for it for the full duration of its execution. This boils down to the requirement that you can never schedule a gate at an earlier time than any gate that has been scheduled already, assuming resources are only allocated by a gate during cycles that it executes.

jvansomeren replied:
I propose to delete the functionality of using resource managing in the scoring of alternatives inside the router.
This means to drop baserc (in favor of base) and minextendrc (in favor of minextend).
Then in FreeCycle the resource managing can be deleted as well.

pablolh replied:
Done

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.cc:
```diff
@@ -127,105 +58,79 @@ utils::Bool FreeCycle::is_first_swap_earliest(
-void FreeCycle::add_no_rc(const ir::compat::GateRef &g, utils::UInt startCycle) {
-    utils::UInt duration = (g->duration+ct-1)/ct;   // rounded-up unsigned integer division
+void FreeCycle::add(const ir::CustomInstructionRef &g, utils::UInt startCycle) {
+    utils::UInt duration = g->instruction_type->duration;
```
Could there be a comment/assert here that this assumes that at startCycle, it indeed can be scheduled?

pablolh replied:
Done

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.cc:
```diff
@@ -127,105 +58,79 @@ utils::Bool FreeCycle::is_first_swap_earliest(
+utils::UInt FreeCycle::cycle_extension(const ir::CustomInstructionRef &g) const {
+    ir::OperandsHelper ops(platform, *g);
+    
+    QL_ASSERT(ops.numberOfQubitOperands() == 1);
```
Why only 1q gates? Where is this cycle_extension function used for?

pablolh replied:
It's only used to know if prepz can be added when choosing prep_z + move
And the implementation is only valid if the gate has one single qubit operand, so, if someone ever needs that for 2q gates they have to change the implementation.

jvansomeren replied:
#479

jvansomeren reviewed (commented):
Reviewed and commented upto and including FreeCycle.*

jvansomeren commented on src/ql/pass/map/qubits/map/detail/future.cc:
```diff
@@ -13,143 +14,205 @@ namespace qubits {
+        // Build DDG and add it as annotation to IR.
+        com::ddg::build(platform, block, options->commute_multi_qubit, options->commute_single_qubit);
+
+        if (options->tie_break_method == TieBreakMethod::CRITICAL) {
```
Is this the option to indicate that you prefer taking the most critical available gate first?

jvansomeren replied:
I would expect that whether the list of gates returned is in criticality order would be represented by a dedicated ddg option.

jvansomeren replied:
It looks like an issue of options to common facilities, common between multiple passes,
where passes receive the options, not the common facility.
How would this be handled?

jvansomeren replied:
#489

jvansomeren commented on src/ql/pass/map/qubits/map/detail/alter.h:
```diff
@@ -21,190 +10,81 @@ namespace qubits {
      */
-    void split(utils::List<Alter> &result) const;
-
+    utils::Bool score_valid = false;
```
Was the policy to know whether the score was valid changed? What was it and what is it now?

pablolh replied:
I am no longer sure, but the API is such that the score is only set at one given point later in the lifetime of this class, before that it is 0/garbage.
This score_valid is only used in an assertion, to check that the score was set before retrieving it.

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.cc:
```diff
@@ -64,58 +33,20 @@ utils::UInt FreeCycle::get_min() const {
-    for (const auto &v : fcv) {
-        if (max_free_cycle < v) {
-            max_free_cycle = v;
+    for (const auto &p : fcv) {
```
This code with its 'p.second' depends on the structure of fcv. 
Some comment at the start of free_cycle.cc summarizing this would be nice. Also the 'p.second' is not very informative, i.e. in the mean time I understand that it is a list; is the index in the list relevant, what is the type of this p.second? Only by that it is compared with max_free_cycle I guess that it is a cycle type. But which one? I understand that sometimes cycles are negative in backward scans in the scheduler; can those negative values be values of cycles here (of course not now because the mapper runs forward but it indicates questions about unclarity of the type here and how future resistant this code is).

pablolh replied:
There is already a comment for fcv:
```c++
     * The map from qubit references to the first cycle index where the given qubit is available.
     * This is encoded as an association list, which avoids the burden of defining a hash or an ordering.
```

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.h:
```diff
@@ -21,151 +16,124 @@ namespace map {
-     */
-    utils::Opt<rmgr::State> rs;
+private:
+    static std::array<utils::UInt, 5> get_indices(const ir::Reference &ref) {
```
This 5, is it an attribute of the ir or of cqasm or of libqasm?

jvansomeren replied:
Can be made part of the IR representation of virtual and real qubits.

jvansomeren replied:
#480

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.h:
```diff
@@ -21,151 +16,124 @@ namespace map {
-     * resource map. This is done because add_no_rc is used to represent just gate
-     * dependencies, avoiding a build of a dep graph.
+     * The map from qubit references to the first cycle index where the given qubit is available.
+     * This is encoded as an association list, which avoids the burden of defining a hash or an ordering.
```
It is a pity that, which qubits general variables, that they don't have an index anymore identifying them that can be used to index in an array here to find the free cycle, i.e. that indexing is replaced by a search. This 'burden' of faster access could make this code more scalable.

jvansomeren replied:
If still a problem, can be made part of the IR representation of virtual and real qubits. Virtual qubits would be similar to qubit variables.

jvansomeren replied:
#480 simplifying IR

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.h:
```diff
@@ -21,151 +16,124 @@ namespace map {
      */
-    void add_no_rc(const ir::compat::GateRef &g, utils::UInt startCycle);
+    std::list<std::pair<ir::Reference, utils::UInt>> fcv;
 
```
Finally, the type!
And it is UInt so not negative.
Would an abstraction towards a 'Cycle' type not be in place?

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.h:
```diff
@@ -21,151 +16,124 @@ namespace map {
+ * This allows to compute, for each routing alternative, what the cost in terms of circuit depth
+ * will be - the goal of the minextend/minextendrc options in this router being to minimize the overall circuit depth.
+ * 
+ * This is also used in the options base and baserc to check whether the operands of a swap should be swapped
```
Probably true but that operands of a swap are swapped, wouldn't improve the score of an alternative in the base[rc] case. This is indeed a consequence of the unclarity of the interdependence of options. But for me, this use when base[rc] can be taken out.

jvansomeren replied:
So: base only counts swaps as metric; minextend counts circuit extension cycles.
Base doesn't need scheduling since it ignores cycles; so also doesn't need FreeCycle.
Minextend would need scheduling to find cycles but that is replaced by the use of FreeCycle.
The purpose of interchanging swap operands is to decrease the cycle extension; the swap always counts as one swap in base.

jvansomeren replied:
#486 past and lists of gates
#488 swap (q1,q2)

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.h:
```diff
@@ -21,151 +16,124 @@ namespace map {
+ * will be - the goal of the minextend/minextendrc options in this router being to minimize the overall circuit depth.
+ * 
+ * This is also used in the options base and baserc to check whether the operands of a swap should be swapped
+ * or whether move should be used instead of swap when allowed.
```
Both move and swap would count as a single operation, so again the difference doesn't show up in the score. Can as well be taken out. I assume that this is a consequence of the scheduling being done during alternative commit, so I would propose to have a templated interface here, one for minextend[rc], one for base[rc], and perhaps later for minfidelity, etc. to implement that scheduling resp. adding it to the list of gates done.

jvansomeren replied:
#486

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.h:
```diff
@@ -21,151 +16,124 @@ namespace map {
      */
-    ir::compat::PlatformRef platform;
+    utils::UInt get_max() const;
 
```
The get_min() is not there. The get_max/get_min would also be used to define the 'output' window: the list of outputted gates in which still newly mapped gates could be scheduled. When the cycle of a gate would drop below the get_min(), that gate would drop out of the window. Only gates in the window need to be taken into consideration during comparing alternatives on minextend. In the old (or older) implementation, the Past contained a copy of the full output list, this copy was copied into every alternative, and thus this slowed down the router for later circuits and during recursion. It was/is not scalable.

jvansomeren replied:
This can be ignored when scheduling is taken out of the computation of the minextend score and replaced by FreeCycle. 
Then the gate lists need not be maintained, so there is no need for a 'window' on it, and so there is no need to keep this window small, hence there is no need to compute get_min().

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.h:
```diff
@@ -21,151 +16,124 @@ namespace map {
+        utils::UInt fr0,
+        utils::UInt fr1,
+        utils::UInt sr0,
+        utils::UInt sr1) const;
```
Still open for me how this could be made configurable.

jvansomeren replied:
#488

jvansomeren commented on src/ql/pass/map/qubits/map/detail/free_cycle.h:
```diff
@@ -21,151 +16,124 @@ namespace map {
-    utils::UInt get_max() const;
+        return result;
+    }
 
```
There are get_for_qubit and get_for_reference, there is a make_qubit_ref, etc.
But I would like to have some comment why these are there.
They are lookup code but why that coding is done in this way, is not clear.

jvansomeren replied:
IR access to find qubit indices in operands is not straightforward.
#480 addresses IR simplification

jvansomeren commented on src/ql/pass/map/qubits/map/detail/future.cc:
```diff
@@ -13,143 +14,205 @@ namespace qubits {
+        ++it;
+    }
+
+    virtual utils::List<ir::CustomInstructionRef> getCurrent() override {
```
That a list is returned, you explained that this is the 'available list', but some comment could do here.

jvansomeren replied:
Part of comment/documentation action.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/future.cc:
```diff
@@ -13,143 +14,205 @@ namespace qubits {
+    }
+
+    virtual void advance(const ir::CustomInstructionRef& gate) override {
+        QL_ASSERT(gate == it->as<ir::CustomInstruction>());
```
Is wait also a custom instruction?

jvansomeren commented on src/ql/pass/map/qubits/map/detail/future.cc:
```diff
@@ -13,143 +14,205 @@ namespace qubits {
+            utils::OutFile(fname.str()).write(dot_graph.str());
+        }
+        
+        const auto& graph = block->get_annotation<com::ddg::Graph>();
```
Comment or reference to comment on the data structure of the ddg would be nice here.

jvansomeren replied:
The use of annotations is the real issue here, not the documentation of the ddg.
So this is part of the IR extension discussion by using annotations,
and thus effectively the discussion how to improve the annotation interface to make it easier to grasp.
Or finally, to document this annotation interface better.

jvansomeren replied:
#481 on annotations
#479 on documentation

jvansomeren commented on src/ql/pass/map/qubits/map/detail/future.h:
'lag'? Is this 'gates'?

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.cc:
```diff
@@ -32,67 +24,25 @@ std::ostream &operator<<(std::ostream &os, PathStrategy p) {
 using namespace utils;
 using namespace com;
 
-/**
```
Again, this comment would be nice to be kept.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.cc:
```diff
@@ -187,147 +105,69 @@ void Mapper::gen_shortest_paths(
-        gen_alters_gate(gate, alters, past);
+List<Alter> Mapper::gen_alters(const utils::List<ir::CustomInstructionRef> &gates, Past &past) {
+    if (options->lookahead_mode != LookaheadMode::ALL) {
+        return gen_alters_gate(gates.front(), past);
```
Is the most-critical one first?
Shouldn't that be somehow here instead of where I read about TIE_BREAK_CRITICAL?

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.cc:
```diff
@@ -374,740 +205,226 @@ Alter Mapper::tie_break_alter(List<Alter> &alters, Future &future) {
-    alter.add_swaps(past, options->swap_selection_mode);
+    alter.add_swaps(past);
 
     // When only some swaps were added (based on configuration), the target
```
By 'configuration' is meant: option 'swap_selection_mode' is 'one'. Only then mapping of a 2q gate may not result in all necessary swaps having been added to make 2q gate NN.

jvansomeren replied:
Is documentation/comment issue.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.cc:
```diff
@@ -374,740 +205,226 @@ Alter Mapper::tie_break_alter(List<Alter> &alters, Future &future) {
-            return false;
-
-        }
+            // FIXME: wait gates as well.
```
I need support to understand this code.
So: the gate is already in ops, so this checks for the current gate to be isNN2QGate (when the option to check for those is true). But I don't understand the argument to this: I would expect 2 qubits to be mapped here and not just 1.

pablolh replied:
the lambda is a v2r mapping applied to both virtual qubit indices to interpret them in the current mapping

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.cc:
```diff
@@ -374,740 +205,226 @@ Alter Mapper::tie_break_alter(List<Alter> &alters, Future &future) {
-        a.debug_print("Considering extension by alternative: ...");
-        a.extend(past, base_past);           // locally here, past will be cloned and kept in alter
-        // and the extension stored into the a.score
+        a.extend(past, base_past); // This fills a.score.
```
I haven't seen past yet. From what you said, I have to assume that both pasts don't contain a list of gates anymore to determine the extension but just free_cycle maps. Would be nice to have some comment here that explains why these two pasts are there and what this extend does.

jvansomeren replied:
Recently reviewing the complexity of the mapper, we found out that 'past' is about to be copied here because the 1st formal parameter of a.extend is not a reference parameter. This copying is only needed to compute the score of the alternative from. The score computation doesn't need the (extended) list of output_gates in the copy, just the FreeCycle map.
So I propose to delete this list of gates copying from this Past copying somehow.
Also as far as I can judge, is the waiting_gates list not used.
The output_gates list is only used for the base_past to store the committed gates in.
This deletion of copying the gates list, reduces the space and time complexity of minextend by one order.

jvansomeren replied:
#486

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.cc:
```diff
@@ -374,740 +205,226 @@ Alter Mapper::tie_break_alter(List<Alter> &alters, Future &future) {
-        }
+    alters.sort([this](const Alter &a1, const Alter &a2) { return a1.get_score() < a2.get_score(); });
+
+    auto factor = options->recursion_width_factor * utils::pow(options->recursion_width_exponent, recursion_depth);
```
These 3 lines are magic. Need comment. So those kept is a fraction (so factor <= 1.0?) of the all alternatives. That stuff in the last line, with those MAX checks, why? Also it is not obvious to me here that recursion_depth increments with the recursing and isn't the maximum depth as according to the option.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.cc:
```diff
@@ -374,740 +205,226 @@ Alter Mapper::tie_break_alter(List<Alter> &alters, Future &future) {
+
+    QL_ASSERT(keep > 0);
+
+    while (alters.size() > keep) {
```
Hopefully the .size is an attribute of alters and need not be computed when asked for. Such attributes are not obvious to me. It requires to know that alters is a list and that this list type has its '.size' implemented as an attribute that is updated on each deletion/addition of an element.

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.cc:
```diff
@@ -374,740 +205,226 @@ Alter Mapper::tie_break_alter(List<Alter> &alters, Future &future) {
-        // goal is to prove that recurse_on_nn_two_qubit should be no at this
-        // place, in the recursion step, but not at level 0!
+
         Bool also_nn_two_qubit_gates = options->recurse_on_nn_two_qubit
```
That option recurse_on_nn_two_qubit is only relevant when lookahead_mode is no_routing_first or all, i.e. not no, could have been coded differently. Or is the option recurse_on_nn_two_qubit also used elsewhere when lookahead_mode may be 'no'?

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.cc:
```diff
@@ -374,740 +205,226 @@ Alter Mapper::tie_break_alter(List<Alter> &alters, Future &future) {
-            a.debug_print(
-                "... ... select_alter, after committing this alternative, mapped easy gates, no gates to evaluate next; RECURSION BOTTOM");
-
+            a.set_score(sub_past.get_max_free_cycle() - base_past.get_max_free_cycle());
```
I'd like to have more insight into when the score is actually computed, i.e. can it be undefined, i.e. is it certain that set_score is always done before get_score? Or does get_score compute it when it hasn't been set?

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.cc:
```diff
@@ -374,740 +205,226 @@ Alter Mapper::tie_break_alter(List<Alter> &alters, Future &future) {
-    //  program's initial mapping for each kernel.
-    
+Mapper::RoutingStatistics Mapper::map_block(ir::BlockBaseRef block) {
     com::map::QubitMapping v2r{
```
Please take care somehow that when initial placement is enabled, that the router must assume one-to-one initialization of v2r. And which are those qubit ids? What is one-to-one?

jvansomeren replied:
This is a comment/documentation issue.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.h:
```diff
@@ -26,20 +17,6 @@ namespace qubits {
-
 /**
  * Strategy options for finding routing paths.
  */
```
Note with LEFT and RIGHT that these strategies only apply to planar topologies, i.e. topologies that can be represented in a plane in such a way that its edges intersect only at their endpoints.

jvansomeren replied:
Documentation/comment issue.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.h:
```diff
@@ -88,43 +62,24 @@ std::ostream &operator<<(std::ostream &os, PathStrategy p);
  *
  * All gates must be unary or two-qubit gates. The operands are virtual qubit
  * indices. After mapping, all virtual qubit operands have been mapped to real
```
Still virtual qubit **indices**?

jvansomeren replied:
Virtual versus real qubit representation issue.

jvansomeren replied:
#480

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.h:
```diff
@@ -88,43 +62,24 @@ std::ostream &operator<<(std::ostream &os, PathStrategy p);
- *    before routing. Which real qubit a virtual qubit maps to is accessed with operator[].
- *    The reverse operation also exists (get_virtual()).
- *    At any time, the virtual to real and reverse maps are 1-1 for all qubits.
+ *    before routing.
```
Indices?!
This is still the case, see com/map/qubit_mapping.h.
So this doesn't fit cqasm1.2 and its IR so needs updating. We have to discuss virtual vs real anyhow.

jvansomeren replied:
Virtual vs real qubit representation issue.

jvansomeren replied:
#480

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.h:
```diff
@@ -134,108 +89,72 @@ std::ostream &operator<<(std::ostream &os, PathStrategy p);
-
     /**
      * Random-number generator for the "random" tie-breaking option.
+     * The seed is constant so that the output of OpenQL is deterministic.
```
So my comments to Medina on nondeterministic behavior in case of random as value to options, is obsolete. Please note.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.h:
```diff
@@ -245,32 +164,15 @@ class Mapper {
-
     /**
      * Find shortest paths between src and tgt in the grid, bounded by a
      * particular strategy. path is a linked-list node representing the complete
```
path from the initial src qubit to **src** in reverse order, not including **src** ?????

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.h:
```diff
@@ -312,50 +213,33 @@ class Mapper {
     );
 
     /**
      * Generates all possible variations of making the given gates
```
Variation => **Alter**s

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.h:
```diff
@@ -312,50 +213,33 @@ class Mapper {
-     * Seeds the random number generator with the current time in microseconds.
+     * nearest-neighbor, starting from given past (with its mappings).
+     * Depending on the lookahead option, only take the first (most
+     * critical) gate, or take all gates and concatenate alternatives.
```
The first is still the most critical by construction?

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.h:
```diff
@@ -364,122 +248,66 @@ class Mapper {
     /**
      * Commit the given Alter, generating swaps in the past and taking it out
-     * of future when done with it. Depending on configuration, this might not
+     * of future. Depending on configuration, this might not
```
Configuration is meant to be: when the option swap_selection_mode is set to 'one', then the target_gate may not be mappable yet after committing this alternative since only a subset of swaps/moves may have been inserted.

jvansomeren replied:
Comment/documentation issue.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.h:
```diff
@@ -364,122 +248,66 @@ class Mapper {
         Future &future,
         Past &past,
-        utils::List<ir::compat::GateRef> &gates,
         utils::Bool also_nn_two_qubit_gates
```
Explain this 'also_nn_two_qubit_gates': see my comments in mail to Medina.

jvansomeren replied:
Comment/documentation issue.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.h:
```diff
@@ -364,122 +248,66 @@ class Mapper {
-     *  - If BASE[_RC], consider all Alters as equivalent, and thus apply the
-     *    tie-breaking strategy to all.
-     *  - If MIN_EXTEND[_RC], consider Alters with the minimal cycle extension
+     *  - If route_heuristic == "base" or "base_rc", consider all alternatives as equivalent,
```
Is it true that for base and base_rc only paths are generated that have the same length. I.e. for multi-core, can there be longer paths?

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.h:
```diff
@@ -364,122 +248,66 @@ class Mapper {
+     *    and thus apply the tie-breaking strategy to all.
+     *  - If route_heuristic == "minextend" or "minextendrc", prefer alternatives
+     *    with the minimal cycle extension
      *    of the given past (or some factor of that amount, ordered by
```
Not a factor of the gates with the minimal cycle extension but a factor of the alternative gates. When the number of alternatives with minimal cycle extension is less than this factor x number of alternatives, also alternatives with more cycles are considered.

jvansomeren replied:
Comment/documentation issue.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/mapper.h:
```diff
@@ -364,122 +248,66 @@ class Mapper {
+     * Process all gates in future and update past with routing result.
      */
-    void map_gates(Future &future, Past &past, Past &base_past);
+    utils::Any<ir::Statement> route_gates(Future &future, Past &past);
```
I'd like to discuss out to the bottom the use of utils::Any etc.

jvansomeren commented on src/ql/pass/map/qubits/map/detail/options.cc:
```diff
@@ -11,36 +7,25 @@ namespace qubits {
 std::ostream &operator<<(std::ostream &os, LookaheadMode lm) {
     switch (lm) {
         case LookaheadMode::DISABLED:             os << "disabled";             break;
-        case LookaheadMode::ONE_QUBIT_GATE_FIRST: os << "one_qubit_gate_first"; break;
```
Why was this taken out?

pablolh replied:
I need to check what it was doing in the old code.

jvansomeren commented on src/ql/pass/map/qubits/map/detail/options.h:
```diff
@@ -24,13 +20,14 @@ enum class Heuristic {
-     * gate scheduling (and thus criticality) is determined without resource
-     * constraints.
+     * gate scheduling is determined without resource
+     * constraints. Gate scheduling is only used for choosing move vs swap, and
```
As commented earlier, I don't mind when gate scheduling is only used for min_extend and not for base. I.e. when base then not for choosing move vs swap and whether to reverse swap operands.

jvansomeren replied:
#488

jvansomeren commented on src/ql/pass/map/qubits/map/detail/options.h:
```diff
@@ -24,13 +20,14 @@ enum class Heuristic {
 
     /**
-     * Same as BASE, but with resource constraints.
+     * Same as BASE, but scheduling is done with resource constraints.
```
Must discuss whether scheduling in case of base makes sense.

jvansomeren replied:
No, doesn't make sense, so scheduling can be taken out with and without resource constraints.
Only for minextend but use FreeCycle there.
And don't require the output_gates in base_past to  be scheduled right.

jvansomeren commented on src/ql/pass/map/qubits/map/detail/options.h:
```diff
@@ -86,8 +72,8 @@ enum class PathSelectionMode {
+    ALL,
     /**
      * Favor routing along the borders of the rectangle defined by the source
      * and target qubit. Only supported when the qubits are given coordinates in
```
Other type of constraint!!! Not that topology is planar.

jvansomeren replied:
Documentation/comment issue.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/options.h:
```diff
@@ -183,7 +160,7 @@ struct Options {
     /**
      * Controls whether the mapper should assume that each qubit starts out
-     * as zero at the start of each kernel, rather than with an undefined
+     * as zero at the start of the block, rather than with an undefined
```
By 'zero' is meant: '|0>'.

jvansomeren replied:
Documentation/comment issue.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/options.h:
```diff
@@ -278,18 +255,14 @@ struct Options {
      */
     utils::Bool write_dot_graphs = false;
 
+    std::string decomposition_rule_name_pattern = "";
```
Need comment here on meaning of this pattern.

jvansomeren replied:
Documentation/comment issue.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/past.h:
```diff
@@ -168,78 +50,26 @@ class Past {
     /**
-     * Prints the state of the embedded FreeCycle object only when verbosity
-     * is at least debug.
+     * Returns a new gate with given name and qubits, throwing if
```
Add to comment that these qubits are mapped qubits, i.e. real qubit indices, as are the operands of all gates in Past.

jvansomeren replied:
Documentation/comment issue.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/detail/past.h:
```diff
@@ -255,18 +85,12 @@ class Past {
 
     /**
      * Generates a single swap/move with real operands and adds it to the
      * current past's waiting list. Note that the swap/move may be implemented
```
Unclear what 'waiting list' is: the list of gates that are to be 'scheduled' into the list of gates in this past.
Note that this comment also applies to add_move but that interface is in this header file before add_swap.
Also note that this gate being 'commutative' is confusing: propose to drop this detail, just mentioning that swap and move just interchange their operands' states.

jvansomeren replied:
Waiting_gates is not used when scheduling is not used.

jvansomeren replied:
#486

jvansomeren commented on src/ql/pass/map/qubits/map/detail/past.h:
```diff
@@ -281,98 +105,56 @@ class Past {
-     * Mapper after-burner. Used to make primitives of all gates that also have
-     * a config file entry with _prim appended to their name, decomposing it
-     * according to the config file gate decomposition.
+     * Flushes the output gate list to the given circuit.
```
See above, also 'output gate list' is not clear, as well what it means to flush to the 'circuit'. Please mention that (for a reason I forgot) gates are first added to a 'waiting list' before being scheduled into the output list (which is meant to be this past's private extension to the mapped circuit). Later when the alternative with this past is committed, this extended output list is copied to the output circuit of the mapper.
These details are relevant because they are visible in the interface. And they remain relevant since classical instructions bypass some of these lists, when encountered.

jvansomeren replied:
After all other comment about avoiding copying the gates lists in Past, what remains of this conversation here is just the documentation/comment issue.

jvansomeren replied:
#486

jvansomeren commented on src/ql/pass/map/qubits/map/detail/past.cc:
```diff
@@ -16,278 +14,70 @@ namespace qubits {
-    QL_ASSERT(kernel->gates.empty());     // kernelp->c will be used by new_gate to return newly created gates into
-    v2r.resize(                       // v2r initializtion until v2r is imported from context
-        nq,
+    v2r.resize(
```
That qubits are all in one object and thereby have unique indices matches cqasm1.0 and not cqasm1.2. So this and the difference between virtual and real qubit representations in the IR needs to be updated.

jvansomeren replied:
Virtual versus real qubit representation issue.

jvansomeren replied:
#480

jvansomeren commented on src/ql/pass/map/qubits/map/detail/past.cc:
```diff
@@ -16,278 +14,70 @@ namespace qubits {
- */
-void Past::add(const ir::compat::GateRef &gate) {
-    waiting_gates.push_back(gate);
+    output_gates.push_back(gate);
```
So there isn't a waiting list anymore; scheduling was just done. Other instructions (not being custom instructions, i.e. not containing qbits and not requiring mapping) may also be encountered. Those don't get a cycle (the cycle is only for quantum gates), and just would be added to output_gates as they are encountered.

jvansomeren replied:
Indeed.

jvansomeren replied:
#486

jvansomeren commented on src/ql/pass/map/qubits/map/detail/past.cc:
```diff
@@ -297,410 +87,116 @@ utils::Bool Past::is_first_swap_earliest(
-        // initcirc going out of scope here so it gets destroyed.
     }
+
+    auto gname = platform->topology->is_inter_core_hop(r0, r1) ? "tmove" : "move";
```
So, the gate names 'swap', 'move', 'tswap' and 'tmove' are reserved names by the router. This is not configurable (in a config) or by options.

jvansomeren replied:
Turn this into an issue?

jvansomeren replied:
#490

jvansomeren commented on src/ql/pass/map/qubits/map/detail/past.cc:
```diff
@@ -297,410 +87,116 @@ utils::Bool Past::is_first_swap_earliest(
-        const ir::compat::SwapParamaters swap_params {true, (utils::Int) r0, (utils::Int) r1, (utils::Int) v1, (utils::Int) v0};
-        gp->swap_params = swap_params;
+
+    if (options->reverse_swap_if_better && fc.is_qubit_free_before(r0, r1)) {
```
This is depending on the target's implementation of a swap. Must be redesigned.
E.g. when adding a new target, and that target also supports swaps, then we describe somehow the decomposition of swap for the target to primitives; this code then here would be an instance of code checking whether for gates with more than 1 operand, one of the operands can start later than other ones. Then with knowledge that a swap allows interchanging its operands and the 'introspection' described above, we can implement this function.

jvansomeren replied:
New issue.

jvansomeren replied:
#490

jvansomeren commented on src/ql/pass/map/qubits/map/detail/past.cc:
```diff
@@ -297,410 +87,116 @@ utils::Bool Past::is_first_swap_earliest(
-        qi = get_real_qubit(qi);          // and now they are real
+void Past::make_real(const ir::CustomInstructionRef &gate) {
+    const auto &gname = gate->instruction_type->name;
+    com::map::ReferenceUpdater::Callback cb = [this, gname](utils::UInt virtual_qubit) {
```
It confuses me that there are virtual/real qubit objects and virtual/real qubit indices. This must be discussed and possibly be the IR must be changed.

jvansomeren replied:
Virtual versus real qubit representation issue.

jvansomeren replied:
#480

jvansomeren commented on src/ql/pass/map/qubits/map/map.cc:
```diff
@@ -91,34 +54,18 @@ void MapQubitsPass::dump_docs(
-      each gate with `_prim`. Thus, if you define a decomposition rule named
-      `cz_prim` rather than `cz`, this rule will only be applied after mapping.
-    )");
+      complexity can be anywhere from linear to exponential!)");
```
Propose to delete this remark or make it precise: in number of gates in the circuit, and in number of qubits in the platform; and for sure in time and perhaps in space.

jvansomeren replied:
Documentation/comment issue.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pass/map/qubits/map/map.cc:
```diff
@@ -127,15 +74,17 @@ MapQubitsPass::MapQubitsPass(
         "Controls whether the mapper should assume that each qubit starts out "
-        "as zero at the start of each kernel, rather than with an undefined "
-        "state."
+        "as zero at the start of the block, rather than with an undefined "
```
zero means |0>
For updates to documentation of this and other options, please see the options overview that I mailed to Medina and Pablo.

jvansomeren replied:
Documentation/comment issue.

jvansomeren replied:
#479

jvansomeren commented on src/ql/pmgr/manager.cc:
```diff
@@ -417,6 +417,10 @@ static utils::Map<utils::Str, utils::Str> convert_global_to_pass_options() {
     if (mapreverseswap.is_set()) {
         retval.set("reverse_swap_if_better") = mapreverseswap.as_str();
     }
+    const auto &mapdecompositionrulepattern = com::options::global["mapdecompositionrulepattern"];
```
This option has an unreadable name, not conforming to e.g. MapDecomp or map_decomp_

jvansomeren replied:
#482

jvansomeren commented on src/ql/rmgr/resource_types/base.cc:
```diff
@@ -137,7 +137,7 @@ utils::Bool Base::gate(
  */
 utils::Bool Base::gate(
-    utils::UInt cycle,
+    utils::Int cycle,
```
Like to have explanation of what happened with backward scheduling resulting in negative cycle values

jvansomeren replied:
#479

jvansomeren commented on src/ql/rmgr/state.cc:
```diff
@@ -63,8 +63,8 @@ utils::Bool State::available(
  * Checks whether the given new-IR statement can be scheduled at the given
  * (start) cycle. Note that the cycle number may be negative.
  */
```
This new name is not really future resistant. Couldn't the old name just be kept and let overloading resolution decide which implementation to take?

jvansomeren commented on src/ql/utils/exception.cc:
```diff
@@ -76,13 +31,12 @@ std::ostream &operator<<(std::ostream &os, ExceptionType etyp) {
     const std::string &msg,
     ExceptionType type
-) noexcept : std::runtime_error(msg), messages({msg}), trace(stack_trace()), type(type) {
+) noexcept : std::runtime_error(msg), messages({msg}), type(type) {
```
As said, I'd like to have those stack traces back.

jvansomeren replied:
Roberto can decide.

jvansomeren commented on tests/golden/test_condex_cnot_last.qasm:
```diff
@@ -1,61 +1,25 @@
+    skip 3
+    measure q[2], b[2]
+    skip 4
+    swap q[3], q[5]
```
Is it intentional that swap is not decomposed?

jvansomeren commented on tests/golden/test_condex_toffoli_composgate_last.qasm:
```diff
@@ -454,7 +233,7 @@ pragma @ql.name("test_condex_toffoli_composgate")
     cond (b[0]) ym90 q[2]
     cond (b[0]) x q[2]
-    { # start at cycle 241
+    { # start at cycle 302
```
The new one is about 60 cycles longer. Why?

jvansomeren replied:
The only change is that there are 17 non-decomposed swaps now versus decomposed ones before. That is about 3 cycles for each swap more cycles on a duration of a swap of 10 cycles.
That could explain it.
Because a non-decomposed swap reserves all resources (and there are several) for the full 10 cycles, while a decomposed swap only does that for the durations of the individual primitives.

jvansomeren replied:
The above was nice to learn.
But the swaps are still in because not decomposed because the decomposer is not run after the router.
#492

jvansomeren commented on tests/golden/test_mapper_allD_last.qasm:
```diff
@@ -1,429 +1,167 @@
+    x q[1]
+    swap q[3], q[1]
+    skip 9
+    cnot q[0], q[3]
```
Not decomposing cnot etc. tests other parts of OpenQL than the original tests did.
So this only tests the router and not decomposition, clifford optimizer, etc.
Still the result seems to be rescheduled.
So there must also be tests that test the router in conjunction with decomposition and clifford optimizer to see that the latter can deal with the output of the router.

jvansomeren replied:
#491

jvansomeren commented on tests/golden/test_mapper_maxcut_last.qasm:
```diff
@@ -1,4 +1,4 @@
-# Generated by OpenQL 0.10.0 for program test_mapper_maxcut
+# Generated by OpenQL 0.10.5 for program test_mapper_maxcut
 version 1.2
```
This seems to be a different circuit. Where are the blue and red gates. I see move_init has disappeared to also the options have changed!

jvansomeren replied:
Is a consequence of not decomposing after the router.
#492

jvansomeren commented on tests/test_mapper.py:
```diff
@@ -394,7 +394,7 @@ def test_mapper_lingling5(self):
         v = 'lingling5'
         config = "cc_light.s17"
-        num_qubits = 7
+        num_qubits = 17
```
Has the meaning of 'num_qubits' changed from virtual number to physical number?

pablolh replied:
It's because while routing on s17, higher qubit indices were reached.
num_qubits is the # of physical qubits, as discussed

jvansomeren commented on tests/test_multi_core_4x4_full.json:
```diff
@@ -264,8 +264,8 @@
     "gate_decomposition": {
-        "tswap_real %0,%1": ["preswap %0","teleportswap %0,%1","postswap %1"],
-        "tmove_real %0,%1": ["premove %0","teleportmove %0,%1","postmove %1"],
+        "tswap %0,%1": ["preswap %0","teleportswap %0,%1","postswap %1"],
```
This decomposition wasn't done in the tests leading to the golden ones above.

jvansomeren replied:
Nonsense remark, please ignore.

jvansomeren replied:
#492

jvansomeren reviewed (dismissed):
Summary (from my head):
documentation
exposing details of data structures
tests changed by testing less but where are the deleted stuff tested
stacktrace deleted
etc.
Probably better to categorize remarks into classes and make an issue for each class.

jvansomeren commented:
We first have to filter the issues out of these conversations before closing this pull request, since they are lost otherwise.
I started by creating the issues. Still must fill them with the applicable conversations above and then resolve those latter.

pablolh commented:
New change is coming up soon

pablolh reviewed (commented):
New change coming soon

Commit:
Fix test_unitary.py with new simulator API

- create C++ version of decomposed_U tests for debugging
- this test is flaky and skipped in CI
- skip also the Python unitary tests in CI

jvansomeren commented on src/ql/pass/map/qubits/map/map.cc:
```diff
@@ -15,72 +11,39 @@ namespace map {
-      definition (i.e. independent on qubit operands), and `cz_real` as a set of
-      specialized gates as required by the platform.
-
-      NOTE: the resolution order is `*_prim`, `*_real`, and finally just the
```
#483 spec/gen
#482 decomposition

jvansomeren commented on tests/cc/test_cfg_none_simple.json:
```diff
@@ -0,0 +1,26 @@
+   "hardware_settings": {
+	 "qubit_number": 17,
+	 "cycle_time" : 20,
+	 "mw_mw_buffer": 0,
```
mw_mw_buffer and all remaining ones until } are not used anymore, can be deleted, as far as I know

jvansomeren commented on tests/golden/test_mapper_lingling7_last.qasm:
```diff
@@ -1,354 +1,186 @@
     }
-    { # start at cycle 285
-        measure q[2]
+    { # start at cycle 315
```
10% worse!
Need to analyze why.
Can also be postponed, but nevertheless ...

jvansomeren replied:
#493

jvansomeren commented on tests/golden/test_mapper_allD_last.qasm:
```diff
@@ -1,432 +1,165 @@
+    skip 3
+    cnot q[6], q[3]
+    skip 3
+    { # start at cycle 310
```
Really bad, from around 200 to around 300 cycles.
Are different options given?
Must analyze!
Is only difference really that mapper's gates are ordered by criticality?

jvansomeren replied:
#493

jvansomeren commented on tests/golden/test_mapper_allNN_last.qasm:
```diff
@@ -1,82 +1,59 @@
+        cnot q[6], q[4]
     }
-    { # start at cycle 27
+    skip 3
```
As written elsewhere, difference also includes not decomposing cnot. This makes resource use overly conservative and thus the circuit more sequential.
That can explain the 10 cycles increase.

jvansomeren replied:
#493
